### PR TITLE
BREAKING(ext/ffi): specialized `buffer` type

### PR DESCRIPTION
**Pull request by littledivy**
This patch splits the `pointer` type into:
- `pointer`: bigint | number | null
- `buffer`: TypedArray | null

`buffer` as return type is the same as `pointer`
 
This is to ensure that users take the fast path in the most common cases. Buffer is optimized for typedarray and pointer is optimized for numbers.

### CODE REVIEW

File: ext/ffi/lib.rs [hunk 1]
```diff
@@ -1059,14 +1066,37 @@ fn ffi_parse_pointer_arg(
...
-  // 4. ArrayBuffer: Fairly common and not supported by Fast API.
+  // 1. BigInt: Uncommon and not supported by Fast API, optimise this case.
+  // 2. Number: Common and supported by Fast API.
+  // 3. Null: Very uncommon / can be represented by a 0.
+  let pointer = if let Ok(value) = v8::Local::<v8::BigInt>::try_from(arg) {
+    value.u64_value().0 as usize as *const u8
+  } else if let Ok(value) = v8::Local::<v8::Number>::try_from(arg) {
+    value.integer_value(scope).unwrap() as usize as *const u8
+  } else if arg.is_null() {
+    ptr::null()
```
Comment (ry) [inline ext/ffi/lib.rs:1077]: Why do the uncommon one first? Seems branch prediction would work better if you reverse this and check for number and null first.

Comment (littledivy) [reply ext/ffi/lib.rs:1077]: numbers enter the fast API call when optimized. This is the slow call so its more likely that the argument is a BigInt or null (least common)

**Comment by aapoalas** [comment]
Should "buffer" be essentially just an alias for "pointer" with the only real difference being how the Fast API is generated? ie. Both would accept numbers and buffers in slow calls, only the Fast API parameter type would change.

File: ext/ffi/jit_trampoline.rs [hunk 2]
```diff
@@ -217,7 +217,7 @@ mod tests {
       }\n\n",
     );
     assert_codegen(
-      codegen(vec![NativeType::Pointer, NativeType::Pointer], NativeType::U64),
+      codegen(vec![NativeType::Buffer, NativeType::Buffer], NativeType::U64),
```
Comment (ry) [inline ext/ffi/jit_trampoline.rs:220]: There are now no tests that exercise a `NativeType::Pointer` argument - could you add some?

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**